kimchi/src/circuits/polynomials/foreign_mul/constraints.rs

```diff
@@ -0,0 +1,40 @@
+//!
+//! Fundamental constraints and related helpers
+//!
+
+use std::cell::RefCell;
+
+use ark_ff::{FftField, One, Zero};
+
+use crate::circuits::expr::{Cache, ConstantExpr, Expr, E};
+
+thread_local! {
+    static CACHE: std::cell::RefCell<Cache>  = RefCell::new(Cache::default());
+}
+
+pub(crate) fn _cache<F: FftField>(mut x: E<F>) -> E<F> {
```

Comment:
Can the cache be passed as an argument (to the `constraints` function of the `Argument` trait) instead of using this cache here?

Reply:
I was trying to avoid passing it around everywhere, but sure can do.  However, it's not being used, so I will just remove it until I actually need it for something.